### PR TITLE
First draft of observability requirements

### LINKED ISSUE #9: Define requirements for Logging Handler
Is this just logging? Or does it handle tracing and metrics too?

Should the SDKs define a common logging interface, or should each platform use the defacto logging choice for the platform.

### PR DESCRIPTION
**Pull request by darrelmiller**
fixes #9
closes #20

### CODE REVIEW

File: Observability.md [hunk 1]
```diff
@@ -0,0 +1,30 @@
+# Observability
+
+[Observability](https://lightstep.com/blog/opentelemetry-101-what-is-observability/) generally refers to the ability to understand an application’s performance based on output data. This telemetry data is divided into three types: metrics, tracing and logging.
+
+Microsoft Graph SDKs should provide the ability for application developers to gather data about the calls made via the SDKs. No telemetry data is captured and transmitted to Microsoft beyond what is included in the HTTP request headers.
```
Comment (sebastienlevert) [inline Observability.md:5]: Should the SDK assume Observability will always be used (Basically creating collectors, exporters and spans in all the methods we believe are relevant) and let developers leverage them (or not)?

Comment (darrelmiller) [reply Observability.md:5]: It is up to the app developer to choose their collectors/exporters. By default we assume none are configured.  Our job us just to create spans.

**Comment by baywet** [comment]
calling out a piece of feedback [provided on the java SDK](https://github.com/microsoftgraph/msgraph-sdk-java-core/issues/480) giving more control over which status codes are logged/traced would also help make the log source usable, it should probably be part of the tags.

File: Observability.md [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+  Send Request: { method, uriTemplate, !url }
+    Obtain Token: { optionalClaims }
+    [Redirect Request]: { statusCode }
+    [Retry Request]: { statusCode, retry-after }
+  Handle Response: { statusCode }
+    [Deserialize Response Payload]: { size, type }
+```
+
+Attributes marked with ! should only be captured if the application has indicated the EUII can be captured.
+Error messages should never include EUII unless it is explicitly indicted by the application.
```
Comment (darrelmiller) [inline Observability.md:42]:
```suggestion
Error messages should never include EUII unless it is explicitly indicated by the application.
```

File: Observability.md [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+## What to capture
+
+Open Telemetry spans are nested and can have attributes attached to each span. The following spans should be created for each request with the indicated attributes.
+
+```
+Application Request: { }
+  Convert RequestInformation to Native Request Object: { }
+    [Serialize Request Payload]: { size, type }
+  Send Request: { method, uriTemplate, !url }
+    Obtain Token: { optionalClaims }
```
[the remark this answers is not in the document]
Comment (baywet) [reply Observability.md:34]: MSAL/azure identity might implement their own tracing spans (which will be nested within ours). But Our authentication providers do more work than "just calling azure identity". They check whether is using https, going to the right host, whether CAE was triggered, etc... before calling into the azure identity API surface. All those decision points are crucial to an admin to understand why any given request might have received a 403.

File: Observability.md [hunk 4]
```diff
@@ -0,0 +1,30 @@
+# Observability
+
+[Observability](https://lightstep.com/blog/opentelemetry-101-what-is-observability/) generally refers to the ability to understand an application’s performance based on output data. This telemetry data is divided into three types: metrics, tracing and logging.
+
+Microsoft Graph SDKs should provide the ability for application developers to gather data about the calls made via the SDKs. No telemetry data is captured and transmitted to Microsoft beyond what is included in the HTTP request headers.
+
+## OpenTelemetry
+
+[OpenTelemetry](https://opentelemetry.io) is an industry standard for capturing observability data from applications. The OpenTelemetry project has a set of libraries that contain the interfaces that an application interacts with to capture data. The primary primitive of OpenTelemetry is a span. An application...
```
[the remark this answers is not in the document]
Comment (baywet) [reply Observability.md:9]: We're taking a dependency on the Open Telemetry API library which only defines the core constructs (similar to abstractions in kiota). The application developer can either choose to ignore tracing (the overhead should be minimal in that case), or add the required collector libraries to send the traces where required.

File: Observability.md [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
+## What to capture
+
+Open Telemetry spans are nested and can have attributes attached to each span. The following spans should be created for each request with the indicated attributes.
```
[the remark this answers is not in the document]
Comment (baywet) [reply Observability.md:27]: some elements of response are already in this comment https://github.com/microsoftgraph/msgraph-sdk-design/pull/72#discussion_r999478285

but effectively each span has its own attributes/tags. We'll add the information an admin might care about for the code we own, which will have their own spans. And MSAL/Azure identity will continue to add their information on their spans. No collision. Much nesting :)

File: Observability.md [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+
+```JSONC
+{
+  "ApplicationRequest?": { // this span is owned by the application itself in the calling code that calls the chained API surface or equivalent
+    "SetContentFromParsable?" { // request information or equivalent when a request body is used from a model, also applies to setContentFromScalar and from stream
+      "com.microsoft.kiota.request.type": "Fully qualified name of the content model"
+    },
+    "SendAsync - <URI template>": { // http request adapter or equivalent, also applies to the other overloads like sendPrimitiveAsync etc...
+      "http.response_content_length?": "int value for the response content length",
```
Comment (baywet) [inline Observability.md:43]:
```suggestion
      "http.response.body.size?": "int value for the response content length",
```

File: Observability.md [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+      "http.status_code": "http response status code",
```
Comment (baywet) [inline Observability.md:44]:
```suggestion
      "http.response.status_code": "http response status code",
```

File: Observability.md [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+      "http.flavor": "1.1 or 2.0",
```
Comment (baywet) [inline Observability.md:45]:
```suggestion
      "network.protocol.version": "1.1 or 2 or 3",
```

File: Observability.md [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+      "http.response_content_type?": "value from the response content type header",
```
Comment (baywet) [inline Observability.md:46]: maybe this should now be
`http.response.body.content_type`

File: Observability.md [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+      "http.method": "http method for the request",
```
Comment (baywet) [inline Observability.md:47]:
```suggestion
      "http.request.method": "http method for the request",
```

File: Observability.md [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+      "http.scheme": "http or https",
```
Comment (baywet) [inline Observability.md:48]:
```suggestion
      "url.scheme": "http or https",
```

File: Observability.md [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+      "http.uri!": "URI of the request",
```
Comment (baywet) [inline Observability.md:49]:
```suggestion
      "url.full!": "URI of the request",
```

File: Observability.md [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+      "http.host": "request host",
```
Comment (baywet) [inline Observability.md:50]:
```suggestion
      "server.address": "request host",
```

File: Observability.md [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+      "http.request_content_length?": "int value for the request content length",
```
Comment (baywet) [inline Observability.md:51]:
```suggestion
      "http.request.body.size?": "int value for the request content length",
```

File: Observability.md [hunk 15]
```diff
@@ -0,0 +1,122 @@
...
+      "http.request_content_type?": "value from the response content type header",
```
Comment (baywet) [inline Observability.md:52]: maybe should be `http.request.body.content_type` now?

File: Observability.md [hunk 16]
```diff
@@ -0,0 +1,122 @@
...
+      "http.uri_template": "URI template for the request",
```
Comment (baywet) [inline Observability.md:53]:
```suggestion
      "url.uri_template": "URI template for the request",
```

**Review by baywet — CHANGES_REQUESTED** [review]

File: Observability.md (hunk 9 above)
Comment (baywet) [reply Observability.md:46]: created a separate issue https://github.com/microsoftgraph/msgraph-sdk-design/issues/99

File: Observability.md (hunk 15 above)
Comment (baywet) [reply Observability.md:52]: created a separate issue https://github.com/microsoftgraph/msgraph-sdk-design/issues/99

**Pull request merged**